### ISSUE TITLE
Large value type properties/ref set accessors

### ISSUE THREAD

**Issue by emceelovin** [dotnet/runtime#63111]
This is probably a very very small use case scenarior, which is why it was never taken into consideration, but I think ref property set accessors should be a thing. it's 100% a fact that they will speed up deserialization. We already have ref get accessors, but ref set accessors are not allowed. This issue is natural by design. I wrote a POCO binary serializer, with reference types, fields, and locals, there is no issue. The issue only comes about when you need to set a large value type property. Again, probably a super small use case, and i'm not trying to turn this into an argument, but i never was one to like when things are done the "wrong" way due to a small number of cases. Right is right IMO. In a runtime like dotnet, data of any kind, memory, disk, socket, etc, should only be read or written the exact number of times it needs to be. Anymore or less is inefficient.

Currently, loading and storing field data by address allows perfect reading/writing. Reading/writing anything of a reference type is obviously also by address and a perfect copy. E.g. you can emit instructions to perfectly write to a large value type field that are equivalent to &reference.field = value. With reference types and setting value type fields, you can effectively emit the equivalent of memcpy(&reference.destinationField, &reference.sourceField, sizeof(ValueType)) with ILGenerator. Consequently, with ValueType's, you can achieve memcpy(<addressOfField><addressOfValueType>valueType.destinationField, <addressOfField><addressOfValueType>valueType.sourceField, sizeof(ValueType)) to copy directly memory to memory.

However, with properties, you're only allowed to have a ref get accessor. Which offers ValueType field like copying semantics, so reading a large ValueType property doesn't take a hit. Writing to a large ValueType property does though. Maybe there is some optimization magic in the JITer?

Taking the following definitions into consideration, if i were to emit code to access a property for this, it would be as follows(i'm typing this into here and not an IDE, so bare with me)

class Program
{
    public static void StringTest(string s)
    {
        int i = 0; // breakpoint here
    }
    
    public struct LargeValueType
    {
        public long LongOne;
        public long LongTwo;
        public long LongThree;
    
        public string StringProperty { get; set; } = "Hello";
    }
    
    public class Serializable
    {
            public LargeValueType LargeValueTypeField = new();
    
            public ref LargeValueType LargeValueTypeProperty
            {
                // manual property backing
                get => ref _refLargValueTypeProperty; // this is ok with the current spec
            
                // this is not allowed currently
                ref set => ref _refLargValueTypeProperty = value; // should be allowed. same as ref var _tmp = ref _refLargValueTypeProperty; _tmp = value;
            
                // auto generated
                // one copy of data initially compared to three(two?) direct copy of contents at &value to contents at &_largeValueTypeProperty/k<>__BackingField
                private void set_LargeValueTypeProperty(ref LargeValueType value)
                    =>
                    ref _largeValueTypePropertyk<>__BackingField = value;
    
                private ref LargeValueType get_LargeValueTypeProperty()
                    =>
                    ref _largeValueTypePropertyk<>__BackingField;
            }
    
           private LargeValueType _largeValueType;
    }
}

var local = DeclareLocal(typeof(Serializable));
var propLocal = DeclareLocal(typeof(LargeValueType));

Emit(OpCodes.Newobj, typeof(Serializable).GetConstructors(Type.EmptyTypes))
Emit(OpCodes.Stloc, local)

// this will throw AccessViolationException when accessing the data on the stack
// local.LargeValueTypeProperty.StringProperty
Emit(OpCodes.Ldloc, local)
Emit(OpCodes.Call, typeof(Serializable).GetProperty("LargeValueTypeProperty").GetGetMethod());
Emit(OpCodes.Call, typeof(LargeValueType).GetProperty("StringProperty").GetGetMethod());
EmitCall(OpCodes.Call, typeof(Program).GetMethod("StringTest"), null);

// for the above code to work
// local.LargeValueTypeProperty.StringProperty
var tmp = DeclareLocal(typeof(LargeValueType));

Emit(OpCodes.Ldloc, local)
Emit(OpCodes.Call, typeof(Serializable).GetProperty("LargeValueTypeProperty").GetGetMethod()); // copy 1
-------------------------------------
Emit(OpCodes.Stloc); // copy2
Emit(OpCodes.Ldloc/Ldloca); // copy 3
-------------------------------------
Emit(OpCodes.Call, typeof(LargeValueType).GetProperty("StringProperty").GetGetMethod());
EmitCall(OpCodes.Call, typeof(Program).GetMethod("StringTest"), null);

// this will execute the way it's intended
// local.LargeValueTypeField.StringProperty
Emit(OpCodes.Ldloc, local)
Emit(OpCodes.Ldfld/Ldflda, typeof(Serializable).GetField("LargeValueTypeField"))
Emit(OpCodes.Call, typeof(LargeValueType).GetProperty("StringProperty").GetGetMethod());
EmitCall(OpCodes.Call, typeof(Program).GetMethod("StringTest"), null);

As you can see, using field data is simple and efficient due to being able to use Ldlfd, but with properties, there is something going on in the runtime that i obviously dont know about. Since it's not possible to execute the way i would THINK it would be. it's also inefficient. there are MULTIPLE unnecessary copies. The data on the stack has to be pushed into a local before you can access any properties etc. I'm assuming that, data must be in a Type housing before any object oriented actions, such as property get/set accessors, can take place? Because value type data can be on the stack and supplied to a function call, but it's consturcted into a parameter. Which leads me to believe it has to be constructed first to be used the way i'm trying to?

**Comment by huoyaoyuan** [comment]
Setting an external ref value requires more complex about lifetime tracking. See https://github.com/dotnet/runtime/issues/32060 .

**Comment by emceelovin** [comment]
You might be thinking what i'm saying the wrong way, and i tried to write it as best as possible to NOT come across that way. When I mean a ref setter, I don't mean having an actual ref property. just the setter would be by reference. i meant to delete the ref in front of the property itself, and the 'ref _backingField = refValue' would be a copy. not a reference assignment. Does that make a little more sense? I'm going to edit this now.

**Comment by huoyaoyuan** [comment]
I still don't understand what you mean.

> The goal is to achieve *object.property = value; which is a regular memory -> register -> memory instruction set in x86/64. without this, its memory -> cpu -> memory(stack) -> cpu -> memory(the memory that the reference type refers to which holds the value type)

How?

You can view the compiled asm at [sharplab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcDMACR2DC6A3utmdgA4BOAlgG5jACm2AygPZRMAyYVA5k1bAqAVwDGwbAH0AZjSYAbACYBuUuSxtOPPoOFjJ2AGphFopgAUq7ChrIk05Z9kFSAvAD4Z8pWvsuEEwe3nIKKtju2AzmTOpO5AC+AQFaVEyy2ly8AkIiElIAShnWtpHe6ZlhfvHOqThwACzYALJIABQc2Xp5hlIQAJQBji7kprGlFJHYELVJKQlkWk2tCJ06Ofr5RoPDAc7FspPTswHJi9j1uM0tmO1DFyOjJmYWx1EAdkwA7lm6uQYCtgiNgAB5IaZIBCYRoAGjBAA5pgBWABsAHYEYk5mRznULssbo17nsLgcSjYpp8fn9Nr0gSDwZDoXDESiMVicdhzud0FoINspF1/ls+sQrop2B9+GCkPDQQh5Zh5azQcj5aj5ej5Qj4okgA). Assigning with regular `set` accessor or `ref` accessor generates the exact same instructions.

**Comment by huoyaoyuan** [comment]
https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcDMACR2DC6A3utmdgA4BOAlgG5jACm2AygPZRMAyYVA5k1bAqAVwDGwbAH0AZjSYAbACYBuUuSxtOPPoOFjJ2AGphFopgAUq7ChrIk05Z9kFSAvAD4Z8pWvsuEEwe3nIKKtju2AzmTOpO5AC+AQFaVEyy2ly8AkIiElIAShnWtpHe6ZlhfvHOqThwACzYALJIABQc2Xp5hlIQAJQBji7kprGlFJHYELVJKQlkWk2tCJ06Ofr5RoPDAc7FspPTswHJi9j1uM0tmO1DFyOjJmYWx1EAdkwA7lm6uQYCtgiNgAB5IaZIBCYRoAGjBAA5pgBWABsAHYEYk5mRznULssbo17nsLgcSjYpp8fn9Nr0gSDwZDoXDESiMVicdg8eQrisWsj2vhsOxgAALJhUB7OJ6jcZvSnTUUSqgAOnlVkpXJ5ZD5N1RQpF4sl0vIspch3eRpVqstWrO6HO6C0EG2Ui6/y2fWIV0U7A+/DBSHhoIQIcwIdZoORIdRIfRIYR8USQA=

Well, when you are copying between objects, the `ref` accessor does win. You can already achieve this with C#. There's just no shorthand syntax for it right now.

**Comment by emceelovin** [comment]
We're starting to get onto the same page here :), and like I mentioned, this is only an issue with large value types.

> I still don't understand what you mean.
> 
> > The goal is to achieve *object.property = value; which is a regular memory -> register -> memory instruction set in x86/64. without this, its memory -> cpu -> memory(stack) -> cpu -> memory(the memory that the reference type refers to which holds the value type)
> 
> How?
> 
> You can view the compiled asm at [sharplab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcDMACR2DC6A3utmdgA4BOAlgG5jACm2AygPZRMAyYVA5k1bAqAVwDGwbAH0AZjSYAbACYBuUuSxtOPPoOFjJ2AGphFopgAUq7ChrIk05Z9kFSAvAD4Z8pWvsuEEwe3nIKKtju2AzmTOpO5AC+AQFaVEyy2ly8AkIiElIAShnWtpHe6ZlhfvHOqThwACzYALJIABQc2Xp5hlIQAJQBji7kprGlFJHYELVJKQlkWk2tCJ06Ofr5RoPDAc7FspPTswHJi9j1uM0tmO1DFyOjJmYWx1EAdkwA7lm6uQYCtgiNgAB5IaZIBCYRoAGjBAA5pgBWABsAHYEYk5mRznULssbo17nsLgcSjYpp8fn9Nr0gSDwZDoXDESiMVicdhzud0FoINspF1/ls+sQrop2B9+GCkPDQQh5Zh5azQcj5aj5ej5Qj4okgA). Assigning with regular `set` accessor or `ref` accessor generates the exact same instructions.

M5 and M6 are exactly what I was trying to get across. Assuming you had a reference that contains a very large value type, which you defined perfectly. Look at how much faster it is if you could load the property by reference just like a field for a copy. It has to be copied onto the stack, then copied from the stack to the destination. Which you can see via ESP usage. This is what i meant by "memory -> register -> memory(stack) -> register -> memory." you can see how it loads the contents of the value type onto the stack. Essentially, the value type was copied for no reason via M5. with M6, and a property ref set accessor, you avoid four copies to the stack. via SIMD instructions, two long's can be copied at a time which is insane performance, but in M5, instead of just copying two long's into a register, and then to a memory address, they're loaded into a register, copied to memory, and then loaded back from memory, to another section of memory. M6 cleans that all up as you can see.

**Comment by emceelovin** [comment]
> https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcDMACR2DC6A3utmdgA4BOAlgG5jACm2AygPZRMAyYVA5k1bAqAVwDGwbAH0AZjSYAbACYBuUuSxtOPPoOFjJ2AGphFopgAUq7ChrIk05Z9kFSAvAD4Z8pWvsuEEwe3nIKKtju2AzmTOpO5AC+AQFaVEyy2ly8AkIiElIAShnWtpHe6ZlhfvHOqThwACzYALJIABQc2Xp5hlIQAJQBji7kprGlFJHYELVJKQlkWk2tCJ06Ofr5RoPDAc7FspPTswHJi9j1uM0tmO1DFyOjJmYWx1EAdkwA7lm6uQYCtgiNgAB5IaZIBCYRoAGjBAA5pgBWABsAHYEYk5mRznULssbo17nsLgcSjYpp8fn9Nr0gSDwZDoXDESiMVicdg8eQrisWsj2vhsOxgAALJhUB7OJ6jcZvSnTUUSqgAOnlVkpXJ5ZD5N1RQpF4sl0vIspch3eRpVqstWrO6HO6C0EG2Ui6/y2fWIV0U7A+/DBSHhoIQIcwIdZoORIdRIfRIYR8USQA=
> 
> Well, when you are copying between objects, the `ref` accessor does win. You can already achieve this with C#. There's just no shorthand syntax for it right now.

Absolutely. The issue is, not with properties, but it's 100% possible with the runtime and JIT. there just needs to be a ref property set accessor that would allow semantics like M6 to make the function literally double in speed on a hot code path.

**Comment by KalleOlaviNiemitalo** [comment]
If the property has a (non-readonly) ref get accessor, then surely callers can use it to get a reference and assign the field via that reference. Why would you then want a ref set accessor as well?

**Comment by emceelovin** [comment]
> If the property has a (non-readonly) ref get accessor, then surely callers can use it to get a reference and assign the field via that reference. Why would you then want a ref set accessor as well?

@KalleOlaviNiemitalo auto implemented properties can't return by reference. Also, ref Identifier Name { get; set;} would imply a ref Identifier hidden/manually generated field. Which isn't the goal here.

**Comment by emceelovin** [comment]
I know what you're saying, but to be syntactically correct about what is what, I think the ref syntax needs to be allowed on the get/set accessor syntax. Since it's 100% legal to have a pointer to a pointer. Would you ever see a ref Identifier Name { ref get; ... }? I highly doubt it, but its possible to point to a pointer.

**Comment by emceelovin** [comment]
private int _backingInt;

public ref int Prop
{
    get => ref _backingInt;
}

@KalleOlaviNiemitalo this would indeed solve half the problem. The reading part, but properties can compute and do other things with a source before they set the internal manually generated backing value, so the writing portion is broken. you can still do the same here, you would just have to do ALL of that computation everywhere you invoked the syntax that invokes the get accessor.

**Comment by emceelovin** [comment]
One of the most important thing's I can think of, is Vector3D. It contains three long properties. Vectors in Direct3D are used a LOT. I'm not an expert in direct3d, but what if you had the following scenario(shooting from the hip because i havent messed with direct3d in forever):

public Vector3D Position { ... }

with public Vector3D Position { ref get; ref set; }

Wouldn't this eliminate the void TranslatePosition(ref Vector3D translation) and allow Player.Position += TranslationVector ?

**Comment by KalleOlaviNiemitalo** [comment]
> Since it's 100% legal to have a pointer to a pointer.

An unmanaged pointer to an unmanaged pointer (C# `byte**`) is OK. A managed pointer to a managed pointer (C# might be `ref ref byte`) is disallowed by ECMA-335 6th edition sections II.14.4.2 and II.23.2.10, and it is not currently permitted by [ECMA-335 CLI Specification Addendum](https://github.com/dotnet/runtime/blob/7ab969c84ef05ba948c0075392716ce335b47744/docs/design/specs/Ecma-335-Augments.md), either.

**Comment by emceelovin** [comment]
> > Since it's 100% legal to have a pointer to a pointer.
> 
> An unmanaged pointer to an unmanaged pointer (C# `byte**`) is OK. A managed pointer to a managed pointer (C# might be `ref ref byte`) is disallowed by ECMA-335 6th edition sections II.14.4.2 and II.23.2.10, and it is not currently permitted by [ECMA-335 CLI Specification Addendum](https://github.com/dotnet/runtime/blob/7ab969c84ef05ba948c0075392716ce335b47744/docs/design/specs/Ecma-335-Augments.md), either.

Fair enough then! You saw where I was coming from.

public ref Identifier Property { get; }

would imply, to me at least, both manually/auto generated:

public ref Identifier Property { get => _manualbackingField/Propertyk<>__BackingField; }

private ref Identifier _manualBackingField/Propertyk<>__BackingField;

**Comment by emceelovin** [comment]
ref fields are only allowed on ref structs also, but that's all irrelevant and an extremely rare circumstance.

**Comment by emceelovin** [comment]
@huoyaoyuan I ran M5 and M6 in a loop with a count of 10 million, inside a loop of 10 iterations, and these were the results:

fastest M5: 223788 ticks(~22.3ms)
fastest M6: 112300 ticks(~11.23ms)

This is just for one property also. the performance is clearly double for large value type's, and if someone was serializing/deserializing a lot of them on a hot path, a 100% performance increase would be valued heavily.

**Comment by JulieLeeMSFT** [comment]
cc @dotnet/jit-contrib .

**Comment by JulieLeeMSFT** [comment]
CC @TIHan.

**Comment by BreyerW** [comment]
Another scenario: raising events inside setter while preserving perf wins due to avoiding copy on value types.

Extra scenario: updating multiple fields without copying large struct. Eg. Imagine object with model matrix and vector3 postion. We might want to update both matrix and position with single matrix4x4 without copying it which is fairly large struct (16 float fields)

**Comment by emceelovin** [comment]
> Another scenario: raising events inside setter while preserving perf wins due to avoiding copy on value types.
> 
> Extra scenario: updating conditionally multiple fields without copying large struct. Eg. Imagine object with model matrix and vector3 postion, rotation, scale. We might want to update both matrix and other mentioned fields depending on what is inside the passed matrix4x4 without copying it which is fairly large struct (16 float fields)

Exactly. Thanks appreciate the added input as well, and I think I even mentioned Direct3D a while ago since all of the translation and what not deals with matrices and vectors. Microsoft.DirectX.Matrix has 16 Single(float) fields. That's 64 bytes per property worth of copy every time a Matrix just happens to be a property somewhere. If it's a field, you can pass it around by reference to whatever uses it in the D3D world that takes references. I'm not really versed on the D3D API though, but once something is a property, you lose all forms of optimization via by ref. There are probably millions of matrices being tossed around in a D3D frame/physics simulation tick/frame. Maybe less. That was just a really random from the hip number, but some games get 300+ FPS. Anything above the simulation tickrate is a "useless" frame in terms of simulation calculation, but i'm sure there are still a TON of transformation matrices being tossed around per tick. Most physics simulations are 64tick CS:GO is 128tick. If someone can even go over all of my speculation about Direct3D, that would be really great. I am by no means an expert.

This is a "problem" that scales infinitely in the negative direction. I mean, I'll be practical here, nobody is gonna put a million fields into a ValueType, but that the amount of fields/properties inside of things has no limit, and it's a 1000000% guarantee/fact that touching that property on a hot path would be atrocious. Serialization and servers are where I see a benefit to this, and ASP dotnet core is ALWAYS looking for speed improvements.

It's just slightly bothersome that I'm not arguing or presenting an opinion, but am trying to help with facts. Problems that never existed in C/C++ due to the nature of the language. There's not a single point in C/C++ where you CANNOT avoid a copy. Over the whole course of csharp's lifetime, the community has been contributing to achieve the same thing and fighting these poor scenarios where unnecessary reading/writing was going on. There's always an update to remove some silly restriction around what csharp prevents that should never have been prevented. ArraySegment was huge, but Span/ReadOnlySpan were absolutely iconic IMO. Especially with stackalloc. The Unsafe class in CompilerServices is another one of my favorites. Maybe even more useful and important than Span/ReadOnlySpan. You can pass around unmanaged ValueType's as a raw chunk of data themselves. You can read/write unmanaged ValueTypes in one shot to a buffer by obtaining a ref byte to them via Unsafe.As<TFrom, TTo> for an EXTREME performance increase that scales infinitely(based on the size of the ValueType) for serialization or whatever. This was as simple as write(&struct) or write(objectPtr) in c++. You always had direct access to memory, and I know that csharp is all about security and code safety, but these additions later on should've been thought of from day one. IMO, csharp is the best programming language currently, and my absolute favorite, but it wasn't always this amazing. csharp 8 didn't become attractive until around csharp 8.0. It suffered for YEARS, and there are still little sharp edges that need to be filed off, like what my OP talked about. I remember when it first came out, I tried c#/WinForms and was just disgusted. It was pure cancer. I went back to c/c++ and good ol win32(rip). GUI programming was the only practical use for it. It was so slow, and the framework lacked heavily, but I'm all for contribution in anyway I can help. I like what's best for the people and the community. I don't mean to sound brash, but anytime a restriction is going to be imposed in the language due to code safety etc, the opposite thought should come into play. How can we safely allow what we just or are about to "take away?"

**Comment by emceelovin** [comment]
> I skimmed the thread. Is this still about wanting additional C# syntax? If so it should be transferred to the csharplang repo.

Yeah, additional syntax is pretty much the only thing. I thought I checked out that repo, but I can't remember. I'm not sure what happened, but I try to put things in the right spot as much as possible. If someone can transfer this to the right spot, that would be great!

**Comment by emceelovin** [comment]
@stephentoub I think the reason I put it into the runtime repo, was due to the CIL I mentioned up there. that's also another issue I had, so is it possible it can be in both repos, or should I separate it?
